openssl/src/pkcs7.rs

```diff
@@ -279,17 +392,191 @@ impl Pkcs7Ref {
+    /// AQ8AMIIBCgKCAQEAt/0hnw55JtLXN4QjFhmkZ5mSSWQM3Zlz8+n9wq99jkDXEUJ8\n\
+    /// YGoHxpBTb9vz4BXwVcR/3GWtdJ/h5VQinBYcZLwtz5iel4IC7pl40a8Kcco3hm6U\n\
+    /// +qve9wz2YS8coQ1+zQ/pqKxDOLN60BYVkuxZeC4yrg8ovL5YftKzeVmFUjDJ/vdI\n\
+    /// RyDScFpso2UoW6mklW/C94ciJH6O9m9dd2nWow1vUfHJlAEm8nPRxJRQH1bROxmc\n\
+    /// 5hLHMjY5wFD/jo31jzxSeZxNhLTKDwn8nG1AGoHaudUNtAF25tzcU6nLKt6BIS+g\n\
+    /// wqavL3kwp+1O/GwLPQb4xXmPs2+f2M08XMbVKwIDAQABMA0GCSqGSIb3DQEBCwUA\n\
+    /// A4IBAQCKWDHBLlLPDmb1C5FOcJ/wqdwCzkbZEBs3qsiim5EDkt9+nqAfwyn0K+G8\n\
+    /// BLJU+6kgeGW1Z0t3RqJdAq2r+7bfMVF8ubJ4zEu5xJAz9UppI4XZ8sZ8iS5rZ3hq\n\
+    /// zjU+7G8Lnu9gEh18q5foi59Wx0jjMyIOWh8O9j3P0JjLxAR8v4rKlYp89/A+vWfb\n\
+    /// TAj31LhWWTa0kiDP9Wd8cMWCjurv7Wq7U4K9gHMHpmUclxs1ByHtFdd61OXSdfBx\n\
+    /// P1Te4tjxQpVo5zURkwOfZoOC6ikAOYoTAYNHP/qwsX2+KnL/JeGCcU4WtoWIp8mt\n\
+    /// HkHgG5EBGoJmLFNFeQWB2yaqAhfD\n\
+    /// -----END CERTIFICATE-----".as_bytes()).unwrap();
+    /// let rsa = Rsa::generate(2048).unwrap();
+    /// let signer_key = PKey::from_rsa(rsa).unwrap();
+    /// let signer_key_clone = signer_key.clone();
+    /// let pkcs7 = Pkcs7::new().unwrap();
+    /// let signer_info = &pkcs7.add_signature(
+    ///     &cert,
+    ///     signer_key_clone,
+    ///     MessageDigest::sha256()
+    /// );
+    ///  ```
+    ///
+    #[corresponds(PKCS7_add_signature)]
+    pub fn add_signature<PT>(
+        &self,
+        cert: &X509Ref,
+        pkey: PKey<PT>,
+        algorithm: MessageDigest,
+    ) -> Result<Pkcs7SignerInfo, ErrorStack>
+    where
+        PT: HasPrivate,
+    {
+        unsafe {
+            let signer_info = cvt_p(ffi::PKCS7_add_signature(
+                self.as_ptr(),
+                cert.as_ptr(),
+                pkey.as_ptr(),
+                algorithm.as_ptr(),
+            ));
+            mem::forget(pkey);
+            signer_info.map(Pkcs7SignerInfo)
+        }
+    }
+
+    /// Add the payload to a PKCS#7 structure.
+    /// The PKCS#7 structure must be either of type NID_pkcs7_signed or NID_pkcs7_digest.
+    /// Finalize a PKCS#7 structure. If the structure's type is `Nid::PKCS7_SIGNED` or
+    /// `Nid::PKCS7_SIGNEDANDENVELOPED`, it will be signed.
+    ///
+    /// The `content_type` must be a PKCS#7 typeAllowed values are
+    /// - Nid::PKCS7_SIGNED
+    /// - Nid::PKCS7_DATA
+    /// - Nid::PKCS7_SIGNEDANDENVELOPED
+    /// - Nid::PKCS7_ENVELOPED
+    /// - Nid::PKCS7_ENCRYPTED
+    /// - Nid::PKCS7_DIGEST
+    /// `content` is the payload of type `content_type`.
+    ///
+    pub fn add_content(&self, content_type: Nid, content: &[u8]) -> Result<MemBio, ErrorStack> {
+        unsafe {
+            // Initialize content
+            cvt(ffi::PKCS7_content_new(self.as_ptr(), content_type.as_raw())).map(|_| ())?;
+            // Write content
+            let bcont_bio = ptr::null_mut();
+            let bio = cvt_p(ffi::PKCS7_dataInit(self.as_ptr(), bcont_bio))
+                .map(|bio| MemBio::from_ptr(bio))?;
```

Comment:
MemBio does not seem like the right type to use here.

Reply:
`PKCS7_dataInit()` returns a `*BIO`. So I think, `MemBio` is exactly what we need here. But you're the expert. Do you have another type in mind? Semantically, I would prefer a reference like `MemBioSlice`, but it's not possible to get a `MemBioSlice` from a pointer.